core/server/worker/src/main/java/alluxio/worker/block/UnderFileSystemBlockStore.java

```diff
@@ -0,0 +1,392 @@
+   * @return the block reader instance
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the
+   * {@link UnderFileSystemBlockStore}
+   * @throws IOException if any I/O errors occur
+   */
+  public BlockReader getBlockReader(final long sessionId, long blockId, long offset,
+      boolean noCache) throws BlockDoesNotExistException, IOException {
+    final BlockInfo blockInfo;
+    mLock.lock();
+    try {
+      blockInfo = getBlockInfo(sessionId, blockId);
+      if (blockInfo.getBlockReader() != null) {
+        return blockInfo.getBlockReader();
+      }
+    } finally {
+      mLock.unlock();
+    }
+    BlockReader reader =
+        UnderFileSystemBlockReader.create(blockInfo.getMeta(), offset, noCache, mAlluxioBlockStore);
+    blockInfo.setBlockReader(reader);
+    return reader;
+  }
+
+  /**
+   * Gets the {@link UnderFileSystemBlockMeta} for a session ID and block ID pair.
+   *
+   * @param sessionId the session ID
+   * @param blockId the block ID
+   * @return the {@link UnderFileSystemBlockMeta} instance
+   * @throws BlockDoesNotExistException if the UFS block does not exist in the
+   * {@link UnderFileSystemBlockStore}
+   */
+  private BlockInfo getBlockInfo(long sessionId, long blockId)
+      throws BlockDoesNotExistException {
+    Key key = new Key(sessionId, blockId);
+    BlockInfo blockInfo = mBlocks.get(key);
+    if (blockInfo == null) {
+      try {
+        throw new BlockDoesNotExistException(ExceptionMessage.UFS_BLOCK_DOES_NOT_EXIST_FOR_SESSION,
+            blockId, sessionId);
+      } catch (Throwable e) {
+        LOG.error("UFS Block does not exist.", e);
+        throw e;
+      }
+    }
+    return blockInfo;
+  }
+
+  private static class Key {
+    private final long mSessionId;
+    private final long mBlockId;
+
+    /**
+     * Creates an instance of the Key class.
+     *
+     * @param sessionId the session ID
+     * @param blockId the block ID
+     */
+    public Key(long sessionId, long blockId) {
+      mSessionId = sessionId;
+      mBlockId = blockId;
+    }
+
+    /**
+     * @return the block ID
+     */
+    public long getBlockId() {
+      return mBlockId;
+    }
+
+    /**
+     * @return the session ID
+     */
+    public long getSessionId() {
+      return mSessionId;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (!(o instanceof Key)) {
+        return false;
+      }
+
+      Key that = (Key) o;
+      return Objects.equal(mBlockId, that.mBlockId) && Objects.equal(mSessionId, that.mSessionId);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(mBlockId, mSessionId);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this).add("blockId", mBlockId).add("sessionId", mSessionId)
+          .toString();
+    }
+  }
+
+  private static class BlockInfo {
```

Comment:
get bit confused by BlockInfo vs. BlockMeta vs ConstMeta, name wise. Also is it necessary to have all of them (can we consolidate one or two)?